File: 0001~0500/0121. Best Time to Buy and Sell Stock/main.cpp
/** 
 * Source: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 * Date: 2022/11/18
 * Skill: 
 * Runtime: 183 ms, faster than 76.33%
 * Memory Usage: 93.6 MB, less than 8.10%
 * Time complexity: 
 * Space complexity: 
 * Constraints: 
 *      
 * 
 */

#include <iostream>
#include <vector>
#include <math.h> 
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min = prices[0], profit = 0;
        for (auto i : prices) {
            if (i - min > profit) profit = i - min;
            if (i < min) min = i;
        }
        return profit;
    }
};

int main() {
    
}

static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0121. Best Time to Buy and Sell Stock/main.py
"""
Source: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/
Date: 2023/2/25
Skill:
Ref:
Runtime: 1003 ms, faster than 82.73%
Memory Usage: 25 MB, less than 32.77%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res, min_value = 0, math.inf
        for price in prices:
            if price < min_value:
                min_value = price
            res = max(res, price - min_value)

        return res


if __name__ == "__main__":
    s = Solution